<a href="https://colab.research.google.com/github/AndrewDisco/didactica2023/blob/main/Teoria_Numerelor.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## În cele ce urmează, aș dori să vă prezint o metodă inovatoare de compunere a problemelor de teoria numerelor folosind programarea. Această metodă combină cunoștințele de matematică cu abilitățile de programare, astfel încât se pot crea probleme interesante și provocatoare.

\\
## Pentru a ilustra această metodă, voi prezenta un program scris în limbajul de programare Python, care găsește perechi de numere ce satisfac o anumită condiție.

In [ ]:
def stringToInt(s: str) -> int:
    result = 0
    for c in s:
        result = result * 10 + (ord(c) - ord('0'))
    return result

def reverses(n: int) -> int:
    s = str(n)
    s = s[::-1]
    return stringToInt(s)

In [ ]:
def findSolutions():
    n = int(input("Cate numere trebuie gasite?\n"))
    count = 0
    seenReverses = set()
    seenPairs = set()
    i = 10
    pairs = set()
    while count < n:
        if i % 10 == 0:
            i += 1
            continue
        square = i * i
        rev = reverses(i)
        if i == rev:
            i += 1
            continue
        revSquare = reverses(square)
        if revSquare == rev * rev and rev not in seenReverses:
            pair1 = str(i) + "^2 = " + str(square)
            pair2 = str(rev) + "^2 = " + str(revSquare)
            if pair1 not in seenPairs and pair2 not in seenPairs:
                pairs.add((pair1, pair2))
                count += 1
                seenReverses.add(rev)
                seenPairs.add(pair1)
                seenPairs.add(pair2)
        i += 1
    return pairs

#Acum funcția poate fi rulată cu ușurință.

In [ ]:
findSolutions()

Cate numere trebuie gasite?
20


{('1002^2 = 1004004', '2001^2 = 4004001'),
 ('1003^2 = 1006009', '3001^2 = 9006001'),
 ('1011^2 = 1022121', '1101^2 = 1212201'),
 ('1012^2 = 1024144', '2101^2 = 4414201'),
 ('1013^2 = 1026169', '3101^2 = 9616201'),
 ('1021^2 = 1042441', '1201^2 = 1442401'),
 ('1022^2 = 1044484', '2201^2 = 4844401'),
 ('102^2 = 10404', '201^2 = 40401'),
 ('1031^2 = 1062961', '1301^2 = 1692601'),
 ('103^2 = 10609', '301^2 = 90601'),
 ('1102^2 = 1214404', '2011^2 = 4044121'),
 ('1103^2 = 1216609', '3011^2 = 9066121'),
 ('1112^2 = 1236544', '2111^2 = 4456321'),
 ('1113^2 = 1238769', '3111^2 = 9678321'),
 ('1121^2 = 1256641', '1211^2 = 1466521'),
 ('112^2 = 12544', '211^2 = 44521'),
 ('113^2 = 12769', '311^2 = 96721'),
 ('122^2 = 14884', '221^2 = 48841'),
 ('12^2 = 144', '21^2 = 441'),
 ('13^2 = 169', '31^2 = 961')}

### Acum, pornind de la problema informatică propusă mai sus și rezultatul funcției rulate, putem elabora o problemă din două părți. Un exemplu ar fi următorul: 
<br></br>

$\textrm{Un număr se numește elocvent dacă acesta îndeplinește următoarele condiții:}$
<br>
$\textrm{1) Acesta are mai mult de o cifră;}$
<br>
$\textrm{2) Inversul pătratului numărului elocvent este egal cu pătratul inversului numărului.}$ <br>
$\textrm{3) Invers}$ $\neq$ $\textrm{Numar elocvent (fără terminațiile în 0-uri)}$
<br></br>

$\textrm{a) Verificați dacă 113 este un număr elocvent.}$ <br>
$\textrm{b) Arătați că există o infinitate de numere elocvente.}$ 
<br> </br>

## Rezolvare

---



$\textrm{b)}$ Pentru a arăta că există o infinitate de numere elocvente, vom construi o familie de astfel de numere.

Fie $n$ un număr natural arbitrar cu cel puțin două cifre. Putem scrie acest număr sub forma $n = \overline{a_1a_2 \cdots a_k}$, unde $a_i$ sunt cifrele lui $n$. Definim numărul $m$ ca fiind $\overline{a_1a_2 \cdots a_ka_ka_{k-1}\cdots a_1}$. Cu alte cuvinte, $m$ este numărul obținut prin concatenarea lui $n$ cu inversul său.

De exemplu, dacă $n = 1234$, atunci $m = 12344321$.

Vom arăta că $m$ este un număr elocvent. Pentru aceasta, trebuie să verificăm cele trei condiții:

1) $m$ are mai mult de o cifră: acest fapt este adevărat deoarece $n$ are cel puțin două cifre.

2) Inversul pătratului numărului $m$ este egal cu pătratul inversului său:

$$\frac{1}{m^2} = \frac{1}{(\overline{a_1a_2 \cdots a_ka_ka_{k-1}\cdots a_1})^2} = \frac{1}{\overline{a_1a_2 \cdots a_k} \cdot \overline{a_ka_{k-1}\cdots a_1}} = \frac{1}{n\cdot \overline{n}} = \left(\frac{1}{n}\right)^2 \cdot \left(\frac{1}{\overline{n}}\right)^2 $$

Deci avem:

$$m^2 = \frac{1}{\left(\frac{1}{m}\right)^2} = \frac{1}{\left(\frac{1}{n}+\frac{1}{\overline{n}}\right)^2} = \frac{n^2 \cdot \overline{n}^2}{(n+\overline{n})^2}$$

Observăm că $n+\overline{n}$ este un număr palindromic, deci putem nota $n+\overline{n} = \overline{b_1b_2 \cdots b_{2k}}$, unde $b_i$ sunt cifre. Atunci:

$$m^2 = \frac{n^2 \cdot \overline{n}^2}{(n+\overline{n})^2} = \frac{n^2 \cdot \overline{n}^2}{\overline{b_1b_2 \cdots b_{2k}}^2} = \frac{n^2 \cdot \overline{n}^2}{\overline{b_{2k}b_{2k-1}\cdots b_2b_1 \cdots b_{2k}b_{2k-1}\cdots b_2b_1}}$$

Deci, inversul pătratului numărului $m$ este egal cu pătratul inversului său:


$$\frac{1}{m^2} = \frac{\overline{b_{2k}b_{2k-1}\cdots b_2b_1}}{n^2 \cdot \overline{n}^2} = \left(\frac{1}{n}\right)^2 \cdot \left(\frac{1}{\overline{n}}\right)^2$$

Astfel, am demonstrat că $m$ este un număr elocvent.

Mai rămâne să verificăm condiția a treia: inversul fără terminații în 0-uri al lui $m$ trebuie să fie diferit de un număr elocvent. Observăm că inversul fără terminații în 0-uri al lui $m$ este $\overline{a_ka_{k-1}\cdots a_1a_k}$, care este diferit de $m$ deoarece are cu două cifre mai puțin și nu poate fi elocvent deoarece are o cifră sau este o putere a lui 10.

Astfel, am arătat că pentru orice număr natural $n$ cu cel puțin două cifre, există un număr elocvent $m$ construit din concatenarea lui $n$ cu inversul său. Deoarece există o infinitate de astfel de numere $n$, există o infinitate de numere elocvente.

In [ ]:
import itertools

def findTriplets(n, m):
    triplets = []
    for a, b, c in itertools.product(range(1, m+1), repeat=3):
        d = int((a**2 + b**2 + c**2)**(1/3))
        if a**2 + b**2 + c**2 == d**3:
            triplets.append((a, b, c))
            if len(triplets) == n:
                break
    return triplets

#Acum funcția poate fi rulată cu ușurință.

In [ ]:
findTriplets(4, 100)

[(1, 1, 5), (1, 5, 1), (3, 3, 3), (5, 1, 1)]

### Avem, așadar o nouă situație în care putem formula o problemă din două părți. Și de această dată, calea ne-a fost facilitată de rularea unei funcții. 
Un exemplu de problemă ar putea fi următorul:<br> </br>

$\textrm{Fie un triplet de numere a, b și c astfel încât } a^2+b^2+c^2=d^3.$ <br>
$\textrm{a) Găsiți tripletul cu cea mai mică sumă a termenilor care îndeplinește cerința.}$ <br>
$\textrm{b) Arătați că există o infinitate de astfel de triplete.}$

## Rezolvare

---
b)
$$
\begin{aligned}
& a^2+b^2+c^2=d^3 \quad / \cdot k^6 ; k \in \mathbb{N} \\
& \left(a \cdot k^3\right)^2+\left(b \cdot k^3\right)^2+\left(c \cdot k^3\right)^2= (d \cdot k^2)^3 \\ \\
& x_1=a \cdot k^3 \\
& x_2=b \cdot k^3 \\
& x_3=c \cdot k^3 \\
& x_4=d \cdot k^2 \\
& \textrm{Cum } k \in \mathbb{N} \textrm{ care poate lua valori infinite, vor exista o infinitate de numere } x_1, x_2, x_3 \textrm{ și } x_4. \\ \\
& \textrm{Astfel, există și o infinitate de numere } a,b,c.
\end{aligned}
$$

In [ ]:
def sum_divisors(n):
    result = 1  # incepem cu 1 deoarece toate numere sunt divizibile cu 1
    for i in range(2, int(n**0.5)+1):
        if n % i == 0:
            result += i
            if i != n//i:
                result += n//i
    if n > 1:
        result += n
    return result

def find_smallest_triplets():
    triplet_found = False
    a = 2
    while not triplet_found:
        sum_a = sum_divisors(a)
        b = a + 1
        while b < sum_a and not triplet_found:
            sum_b = sum_divisors(b)
            c = b + 1
            while c <= sum_b and not triplet_found:
                sum_c = sum_divisors(c)
                if sum_a == sum_b == sum_c:
                    triplet_found = True
                    print("Cele mai mici triplete cu suma divizorilor aceeasi este: ", a, b, c)
                c += 1
            b += 1
        a += 1


In [ ]:
find_smallest_triplets()

Cele mai mici triplete cu suma divizorilor aceeasi este:  14 15 23


### Acum, pornind de la problema informatică propusă mai sus și rezultatul funcției rulate, putem elabora o problemă din două părți. Un exemplu ar fi următorul: 
<br></br>

$\textrm{Fie x,y,z un triplet de numere astfel încât suma divizorilor acestora să fie egală.}$
<br>

$\textrm{a) Găsiți cel mai mic astfel de triplet.}$ <br>
$\textrm{b) Arătați că există o infinitate de triplete de acest fel.}$ 
<br> </br>

## Rezolvare

---

b)
$$
\begin{aligned}
& \left.\begin{array}{l}
(m, n)=1 \\
(m, p)=1 \\
s(n)=s(p)
\end{array}\right\} \implies \quad s(m \cdot n)=s(m \cdot p) \\ \\
& D_n=\left\{d_1^n ; d_2^n ; \ldots ; d_{\tau(n)}^n\right\} \\
& D_p=\left\{d_1^p ; d_2^p ; \ldots ; d_{\tau(p)}^p\right\} \\
& D_m=\left\{d_1^m ; d_2^m ; \ldots  ;d_{\tau(m)}^m\right\} \\
& D_{m \cdot n} = \left\{d_1^n ; d_2^n ; \ldots ;d_{\tau(n)}^n; \\ d_1^m \cdot d_1^n; d_1^m \cdot d_2^n; \ldots \right\} \\ \\
&
\end{aligned}
$$

$$
\begin{aligned}
& \left.\begin{array}{l}
s(m \cdot n)=s(m) \cdot s(n) \\
s\left(m \cdot p\right)=s(m) \cdot s(p) \\
s (n)=s(p)
\end{array}\right\} \implies \quad s(m \cdot n)= s(n \cdot p) \\
&
\end{aligned}
$$

$$
\begin{aligned}
& k=11^n \\
& a_1=14 \cdot 11^n \\
& b_1=15 \cdot 11^n \\
& c_1=23 \cdot 11^n, n \in \mathbb{N}^* \\
& \textrm{Cum } n \textrm{ poate lua valori infinite, și } a_1, b_1, c_1 \textrm{ vor lua un număr infinit de valori.} \\
& \textrm{Astfel, se poate arăta că există un număr infinit de numere cu aceeasi suma a divizorilor.}
\end{aligned}
$$
